datalad/downloaders/shub.py

```diff
@@ -0,0 +1,58 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Support for resolving Singularity Hub URLs
+"""
+
+import json
+from logging import getLogger
+
+from datalad.dochelpers import (
+    borrowkwargs,
+    exc_str,
+)
+from datalad.downloaders.http import HTTPDownloader
+from datalad.support.exceptions import DownloadError
+from datalad.utils import auto_repr
+
+lgr = getLogger("datalad.downloaders.shub")
+
+
+@auto_repr
+class SHubDownloader(HTTPDownloader):
+    """Resolve shub:// URLs before handing them off to HTTPDownloader.
+    """
+
+    api_url = "https://singularity-hub.org/api/container/"
+
+    @borrowkwargs(HTTPDownloader)
+    def __init__(self, *args, **kwargs):
+        self._resolved_urls = {}
+        super().__init__(*args, **kwargs)
+
+    def _resolve_url(self, url):
+        if url in self._resolved_urls:
+            resolved = self._resolved_urls[url]
```

Comment:
Why do we want to cache resolved urls?
 There could be long running sessions in Python during which image could change or resolved URL time out.

Reply:
If we were to cache it, `lru_cache` decorator could probably be used avoiding custom code?

Reply:
Yeah, we probably shouldn't cache here.  In an earlier iteration, the subclass was overriding more methods, and it did need it because the same download/fetch call would trigger it multiple times.  Will remove.

Reply:
FWIW, I was pleasantly surprised by how minimalistic this PR is to accomplish the mission ;-)

Reply:
> FWIW, I was pleasantly surprised by how minimalistic this PR is to accomplish the mission ;-)

Me too.  (As you might imagine, it didn't start out that way.)

Kudos to your downloaders/providers code for providing a single point to hook into.

Reply:
> Kudos to your downloaders/providers code for providing a single point to hook into.

first I wanted to say only "ha -- a single point so hard to find"... but then it also stroke me with what I haven't expressed while looking at the code and thinking to suggest to use delegation more and thus having `_resolve_url` a `@classmethod`. 
Not in this PR but later (when another use case arrives if ever) we might want to RF it so that a specific "redirected" downloader would just be delegating (not subclassing) to another downloader appropriate for the resolved url.  In case of `shub://` ATM we do know that it resolves to an http url, so we can just subclass.  But generally it might not be the case: may be it would resolve to `s3://` (unlikely -- not standard) etc. May be it would also be relevant for the cases when redirected to url would have its providers' configuration to authenticate etc.  So, in principle, delegation would have been  a more generic way to implement it.  For that we could have some base class (e.g. `RedirectedDownloader`) which would provide common logic, and subclasses would only implement `_resolve_url` giving it back to the base class logic to instantiate appropriate for the url downloader etc.